### PR TITLE
Mod 2 (co)homology as a module over the Steenrod algebra

### LINKED ISSUE #6103: cohomology of simplicial complexes with finite coefficients as modules over steenrod algebras
<div id="comment:0"></div>

Implement functionality in sage to return the ZZ-graded cohomology with GF(p)-coefficients of an arbitrary simplicial complex as a module over the Steenrod algebra A(p).

This will depend on #6099, #6100, #6101, and #5882. The methodolgy will be similar to that implemented in the ticket #6102.

CC:  @jhpalmieri @antieau

Component: **algebraic topology**

_Issue created by migration from https://trac.sagemath.org/ticket/6103_

### PR DESCRIPTION
**Pull request by jhpalmieri**
Mod 2 (co)homology as a module over the Steenrod algebra

Implement:

- vector space duality for (co)homology of cell complexes with field coefficients, switching between homology and cohomology
- Steenrod operations on homology and cohomology, acting from the left and from the right

The Steenrod operation "Sq(i)" was already implemented for cohomology rings, raising an error if it was used with a base field other than GF(2). Now there is a new class, `CohomologyRing_mod2`, which should only be used with that base field, and the `Sq` method is attached to Elements of that new class.

Using Steenrod operations is more transparent: in addition to doing `x.Sq(i)`, users can do `Sq(i) * x` or `x * Sq(i)`, or indeed `a * x` or `x * a` for any element `a` of `SteenrodAlgebra(2)`. (The elements `Sq(i)` generate the Steenrod algebra, but they do not constitute a full list of the elements.)

This fixes #6103.

### :memo: Checklist

- [X] The title is concise, informative, and self-explanatory.
- [X] The description explains in detail what this PR is about.
- [X] I have linked a relevant issue or discussion.
- [X] I have created tests covering the changes.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 1]
```diff
@@ -385,6 +393,67 @@ def _to_cycle_on_basis(self, i):
...
+        """
+        okay = True
+        dual = self.dual()
+        dims = [a[0] for a in self._indices]
+        for dim in range(max(dims) + 1):
+            n = len(self.basis(dim))
+            m = matrix(n, n, [a.eval(b) for a in self.basis(dim) for b in dual.basis(dim)])
+            okay = (m == 1)
+            if not okay:
+                print('error in dimension {}'.format(dim))
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
        tester = self._tester(**options))
        dual = self.dual()
        dims = [a[0] for a in self._indices]
        for dim in range(max(max(dims),  tester._max_runs) + 1):
            n = len(self.basis(dim))
            m = matrix(n, n, [a.eval(b) for a in self.basis(dim) for b in dual.basis(dim)])
            tester.assertEqual(m, 1,f "error in dimension {dim}")
```
This is the standard `_test_*` method paradigm. This also allows you to limit the dimensions checked.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 2]
```diff
@@ -413,6 +482,221 @@ def to_cycle(self):
...
+        def eval(self, other):
+            r"""
+            Evaluate ``self`` at ``other``.
+
+            INPUT:
+
+            - ``other`` -- an element of the dual space: if ``self``
+              is an element of cohomology in dimension `n`, then
+              ``other`` should be an element of homology in dimension
+              `n`, and vice versa.
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
            - ``other`` -- an element of the dual space; if ``self``
              is an element of cohomology in dimension `n`, then
              ``other`` should be an element of homology in dimension
              `n`, and vice versa
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 3]
```diff
@@ -413,6 +482,221 @@ def to_cycle(self):
...
+                2
+            """
+            if not self or not other:
+                return self.base_ring().zero()
+            if self.parent()._cohomology:
+                return self.to_cycle().eval(other.to_cycle())
+            else:
+                return other.to_cycle().eval(self.to_cycle())
+
+class HomologyVectorSpaceWithBasis_mod2(HomologyVectorSpaceWithBasis):
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py:524]:
```suggestion


class HomologyVectorSpaceWithBasis_mod2(HomologyVectorSpaceWithBasis):
```
PEP8 thing.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 4]
```diff
@@ -413,6 +482,221 @@ def to_cycle(self):
...
+                sage: (Sq(1) * Sq(2)) * x4 != 0
+                True
+                sage: (Sq(1) * Sq(2)) * x4 == Sq(1) * (Sq(2) * x4)
+                True
+                sage: x4 * (Sq(2) * Sq(1)) == (x4 * Sq(2)) * Sq(1)
+                True
+            """
+            # Handle field elements first.
+            if a in self.base_ring():
+                return self.map_coefficients(lambda c: c*a)
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]: Is there some reason you are not using the default implementation given by CFM elements for scalars?

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 5]
```diff
@@ -413,6 +482,221 @@ def to_cycle(self):
...
+            if not self_on_left: # i.e., module element on left
+                a = a.antipode()
+
+            m = self.degree()
+            n = a.degree()
+            if m <= n:
+                return self.parent().zero()
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py:693]: Does the antipode change the degree? If not, you should put this case before the `self_on_left` check as the antipode computation is unnecessary.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 6]
```diff
@@ -413,6 +482,221 @@ def to_cycle(self):
...
+
+            vec = []
+            for x in sorted(self.parent().dual().basis(m-n)):
+                vec.append(self.eval(a * x))
+            B = list(self.parent().basis(m-n))
+            return self.parent().linear_combination(zip(B, vec))
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
            P = self.parent()
            B = list(P.basis(m-n))
            return P._from_dict({b.support()[0]: self.eval(a * x)
                                 for x in sorted(self.parent().dual().basis(m-n))})
```
This should be equivalent and faster.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 7]
```diff
@@ -676,6 +937,96 @@ def cup_product(self, other):
...
+        Initialize ``self``.
+
+        EXAMPLES::
+
+            sage: RP2 = simplicial_complexes.ProjectivePlane()
+            sage: H = RP2.cohomology_ring(GF(2))
+            sage: TestSuite(H).run()
+        """
+        if base_ring != GF(2):
+            raise ValueError
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
            raise ValueError("the base ring must be GF(2)")
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 8]
```diff
@@ -676,6 +937,96 @@ def cup_product(self, other):
...
+
+        Implement Steenrod operations on (co)homology at odd primes,
+        and thereby implement this class over `\GF{p}` for any `p`.
+
+    INPUT:
+
+    - ``base_ring`` -- must be the field ``GF(2)``
+    - ``cell_complex`` -- the cell complex whose homology we are
+      computing
+    - ``category`` -- (optional) a subcategory of modules with basis
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]: `category` is currently not an input.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 9]
```diff
@@ -841,6 +1208,188 @@ def Sq(self, i):
...
+            TESTS::
+
+                sage: # needs sage.groups
+                sage: (Sq(2) * Sq(1)) * x == Sq(2) * (Sq(1) * x)
+                True
+                sage: x * (Sq(1) * Sq(2)) == (x * Sq(1)) * Sq(2)
+                True
+            """
+            # Handle field elements first.
+            if a in self.base_ring():
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]: Same comment about going to CFMElement.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 10]
```diff
@@ -841,6 +1208,188 @@ def Sq(self, i):
...
+          ring
+
+        - ``deg_codomain`` -- the degree of the codomain in the
+          cohomology ring
+
+        - ``side`` (optional, default ``'left'``) -- are we computing
+          the action as a left module action or a right module? (This
+          documentation is written from the point of view of a left
+          action for brevity. Just switch all of the tensors for the
+          right action.)
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
        - ``side`` default ``'left'``) -- whether we are computing
          the action as a left module action or a right module
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 11]
```diff
@@ -841,6 +1208,188 @@ def Sq(self, i):
...
+
+        Writing `m` for ``deg_domain`` and `n` for ``deg_codomain``, this
+        returns `A^{n-m} \otimes H^{m} \to H^{n}`, one single
+        component of the map making `H` into an `A`-module.
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py:1293]:
```suggestion
        We will write this with respect to the left action;
        for the right action, just switch all of the the tensors.
        Writing `m` for ``deg_domain`` and `n` for ``deg_codomain``, this
        returns `A^{n-m} \otimes H^{m} \to H^{n}`, one single
        component of the map making `H` into an `A`-module.
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 12]
```diff
@@ -841,6 +1208,188 @@ def Sq(self, i):
...
+
+           This is only implemented in characteristic two. The main
+           implementation is only for simplicial complexes and simplicial
+           sets; cubical complexes are converted to simplicial complexes
+           first. Note that this converted complex may be large and so
+           computations may be slow. There is no implementation for
+           `\Delta`-complexes.
+
+        Algorithm: use the Milnor basis for the truncated Steenrod
+        algebra `A`, and for cohomology, use the basis with which it
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
        ALGORITHM:

        Use the Milnor basis for the truncated Steenrod
        algebra `A`, and for cohomology, use the basis with which it
```

File: src/sage/topology/cell_complex.py [hunk 13]
```diff
@@ -865,7 +865,12 @@ def homology_with_basis(self, base_ring=QQ, cohomology=False):
             sage: list(H.basis(3))                                                      # needs sage.modules
             [h^{3,0}]
         """
-        from sage.homology.homology_vector_space_with_basis import HomologyVectorSpaceWithBasis
+        from sage.homology.homology_vector_space_with_basis import \
+            HomologyVectorSpaceWithBasis, HomologyVectorSpaceWithBasis_mod2
+        if base_ring == GF(2):
+            if cohomology:
+                return self.cohomology_ring(base_ring)
```
Comment (tscrim) [inline src/sage/topology/cell_complex.py]:
```suggestion
        if cohomology:
            return self.cohomology_ring(base_ring)
        if base_ring == GF(2):
```
Otherwise it would not match the documentation outside of `GF(2)`.

**Review by tscrim — CHANGES_REQUESTED** [review]

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 4 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py]: I thought I had tried other things and they hadn't worked. For example I thought that just using `return c * self` led to an infinite recursion, because `*` called `_acted_upon_`. This doesn't seem to be the case, so I must have been confused. I'll change it.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 5 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py:693]: Good idea.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 9 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py]: Ah, this is the place where I get the infinite recursion. I think I need to keep this one as is.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 7 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py]: This should never happen — the class should only ever be instantiated when the base ring is GF(2) — but the change is fine.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 8 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py]: Will fix.

File: src/sage/topology/cell_complex.py (hunk 13 above)
Comment (jhpalmieri) [reply src/sage/topology/cell_complex.py]: Right, good catch.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 9 above)
Comment (tscrim) [reply src/sage/homology/homology_vector_space_with_basis.py]: You should be able to do
```python
ret = CombinatorialFreeModule.Element._acted_upon_(self, a, self_on_left)
if ret is not None:  # did the scalar action
    return ret
```
which will be a lot faster than the current implementation. If this doing this does get an infinite loop, can you post the traceback? It really shouldn't have such a loop...

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 14]
```diff
@@ -413,6 +480,221 @@ def to_cycle(self):
...
+                sage: (Sq(1) * Sq(2)) * x4 != 0
+                True
+                sage: (Sq(1) * Sq(2)) * x4 == Sq(1) * (Sq(2) * x4)
+                True
+                sage: x4 * (Sq(2) * Sq(1)) == (x4 * Sq(2)) * Sq(1)
+                True
+            """
+            # Handle field elements first.
+            if a in self.base_ring():
+                return a * self
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]: This wasn't what I had in mind. I was thinking of an explicit `super()` call (or to the specific base class as indicated in the other comment).

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 15]
```diff
@@ -413,6 +480,221 @@ def to_cycle(self):
...
+                \langle (S'' \phi) (f \otimes a), x \rangle
+                = \langle f, \phi_L (a \otimes x) \rangle,
+
+            for `f \in H_m`, `a \in A^n`, and `x \in
+            H^{m-n}`. Somewhat more succintly, we define the action `f
+            \cdot a` by
+
+            .. MATH::
+
+                (f \cdot a) (x) = f (a \cdot x)
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
                (f \cdot a) (x) = f (a \cdot x).
```

**Review by tscrim — CHANGES_REQUESTED** [review]
Two other general comments:

1. Can you please squash all of the little commits from the previous review? They are a bit annoying and the messages are worse than useless.
2. This is agnostic to the implementation of $F_2$, right? So I think we need a more robust test as we have, e.g.,
```python
sage: GF(2) == GF(2, impl='ntl')
False
```
I am thinking of instead testing the characteristic and the cardinality. Actually, is this something for any char 2 field or just $F_2$?

**Comment by jhpalmieri** [comment]
> Two other general comments:
> 
>     1. Can you please squash all of the little commits from the previous review? They are a bit annoying and the messages are worse than useless.

Sure.
 
>     2. This is agnostic to the implementation of F2, right? So I think we need a more robust test as we have, e.g.,
> 
> 
> ```python
> sage: GF(2) == GF(2, impl='ntl')
> False
> ```

This looks like a bug in Sage to me. I can understand why `GF(2) is GF(2, impl='ntl')` might return False, but not `==`. I will work around it.
 
> I am thinking of instead testing the characteristic and the cardinality. Actually, is this something for any char 2 field or just F2?

Just F2.

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 15 above)
Comment (jhpalmieri) [reply src/sage/homology/homology_vector_space_with_basis.py]: Okay

**Comment by tscrim** [comment]
> > Two other general comments:
> > 1. Can you please squash all of the little commits from the previous review? They are a bit annoying and the messages are worse than useless.
> 
> Sure.

Thank you.

> > 2. This is agnostic to the implementation of F2, right? So I think we need a more robust test as we have, e.g.,
> > ```python
> > sage: GF(2) == GF(2, impl='ntl')
> > False
> > ```
> 
> This looks like a bug in Sage to me. I can understand why `GF(2) is GF(2, impl='ntl')` might return False, but not `==`. I will work around it.

It's not a bug, but a feature. Equal-but-not-identical parents are a big pain to deal with, and IMO the less we have the better. I think `isinstance(X, FiniteField) and X.cardinality() == 2` is a sufficiently good test.  Although what you did will work. You should also change the ` CohomologyRing_mod2.__init__` method and add some doctests using other implementations rather than the default.

Also, for the category, you can do `category = Cat.or_subcategory(category)` to catch potential bugs; it also can take `None` as input.

**Comment by jhpalmieri** [comment]
Regarding `GF(2)`, should I be concerned that `SteenrodAlgebra(2)` uses a specific implementation of the field, and it may not be the same as the one here? Computations seem to work.

> It's not a bug, but a feature. Equal-but-not-identical parents are a big pain to deal with, and IMO the less we have the better. I think `isinstance(X, FiniteField) and X.cardinality() == 2` is a sufficiently good test. Although what you did will work. You should also change the `CohomologyRing_mod2.__init__` method 

I'm not sure what changes you have in mind. The `cat.or_subcategory(...)` suggestion below, or something else?

> and add some doctests using other implementations rather than the default.

It's easy to add some computations of cohomology operations, cup products, etc., but I get errors with `TestSuite(...)` if I use other implementations, with the message `TypeError: cannot pickle 'dict_items' object`. (`_test_pickling` fails with `impl='ntl'`, `_test_pickling` and `_test_elements` fail with `impl='givaro'`.) Okay to ignore these?

> Also, for the category, you can do `category = Cat.or_subcategory(category)` to catch potential bugs; it also can take `None` as input.

**Comment by tscrim** [comment]
> Regarding `GF(2)`, should I be concerned that `SteenrodAlgebra(2)` uses a specific implementation of the field, and it may not be the same as the one here? Computations seem to work.

Perhaps, but I don't really know that code very well. It might be best to put some kind of warning or note about it.

> > It's not a bug, but a feature. Equal-but-not-identical parents are a big pain to deal with, and IMO the less we have the better. I think `isinstance(X, FiniteField) and X.cardinality() == 2` is a sufficiently good test. Although what you did will work. You should also change the `CohomologyRing_mod2.__init__` method
> 
> I'm not sure what changes you have in mind. The `cat.or_subcategory(...)` suggestion below, or something else?

I thought I saw a `== GF(2)` there.

> > and add some doctests using other implementations rather than the default.
> 
> It's easy to add some computations of cohomology operations, cup products, etc., but I get errors with `TestSuite(...)` if I use other implementations, with the message `TypeError: cannot pickle 'dict_items' object`. (`_test_pickling` fails with `impl='ntl'`, `_test_pickling` and `_test_elements` fail with `impl='givaro'`.) Okay to ignore these?

Yes, although it looks like you found some bugs that need reporting. I might add these TestSuite calls with skipping those particular tests.

**Comment by jhpalmieri** [comment]
> > Regarding `GF(2)`, should I be concerned that `SteenrodAlgebra(2)` uses a specific implementation of the field, and it may not be the same as the one here? Computations seem to work.
> 
> Perhaps, but I don't really know that code very well. It might be best to put some kind of warning or note about it.
> 
> > > It's not a bug, but a feature. Equal-but-not-identical parents are a big pain to deal with, and IMO the less we have the better. I think `isinstance(X, FiniteField) and X.cardinality() == 2` is a sufficiently good test. Although what you did will work. You should also change the `CohomologyRing_mod2.__init__` method
> > 
> > 
> > I'm not sure what changes you have in mind. The `cat.or_subcategory(...)` suggestion below, or something else?
> 
> I thought I saw a `== GF(2)` there.
> 
> > > and add some doctests using other implementations rather than the default.
> > 
> > 
> > It's easy to add some computations of cohomology operations, cup products, etc., but I get errors with `TestSuite(...)` if I use other implementations, with the message `TypeError: cannot pickle 'dict_items' object`. (`_test_pickling` fails with `impl='ntl'`, `_test_pickling` and `_test_elements` fail with `impl='givaro'`.) Okay to ignore these?
> 
> Yes, although it looks like you found some bugs that need reporting. I might add these TestSuite calls with skipping those particular tests.

Actually it looks like I was wrong about this being agnostic about the implementation of `GF(2)`. Or I wasn't wrong, but some of the old stuff doesn't work with those other implementations. In particular, simplicial complexes don't work well, but simplicial sets do: just let `K` be your favorite simplicial complex and try `K.cohomology_ring(GF(2, impl='ntl'))`. So I propose that we don't add examples with other implementations here, but instead fix the whole thing on another ticket, adding examples then.

**Comment by jhpalmieri** [comment]
> Actually it looks like I was wrong about this being agnostic about the implementation of `GF(2)`. Or I wasn't wrong, but some of the old stuff doesn't work with those other implementations. In particular, simplicial complexes don't work well, but simplicial sets do: just let `K` be your favorite simplicial complex and try `K.cohomology_ring(GF(2, impl='ntl'))`. So I propose that we don't add examples with other implementations here, but instead fix the whole thing on another ticket, adding examples then.

Maybe I should say "your favorite _nonempty_ simplicial complex," because it looks like the empty complex works okay. `simplicial_complexes.Simplex(0).cohomology_ring(GF(2, impl='ntl'))` fails.

**Comment by jhpalmieri** [comment]
> > Actually it looks like I was wrong about this being agnostic about the implementation of `GF(2)`. Or I wasn't wrong, but some of the old stuff doesn't work with those other implementations. In particular, simplicial complexes don't work well, but simplicial sets do: just let `K` be your favorite simplicial complex and try `K.cohomology_ring(GF(2, impl='ntl'))`. So I propose that we don't add examples with other implementations here, but instead fix the whole thing on another ticket, adding examples then.
> 
> Maybe I should say "your favorite _nonempty_ simplicial complex," because it looks like the empty complex works okay. `simplicial_complexes.Simplex(0).cohomology_ring(GF(2, impl='ntl'))` fails.

Actually, now I really need to be persuaded that `GF(2, impl='ntl')` should not just be thrown in the trash heap:

```
sage: F = GF(2, impl='ntl')
sage: m_ntl = identity_matrix(1, F)
sage: v_ntl = vector(F, (1,))
sage: m_ntl * v_ntl
---------------------------------------------------------------------------
SignalError                               Traceback (most recent call last)
Cell In [14], line 1
----> 1 m_ntl * v_ntl

File ~/Desktop/Sage/git/sage/src/sage/structure/element.pyx:4099, in sage.structure.element.Matrix.__mul__()
   4097 
   4098         if BOTH_ARE_ELEMENT(cl):
-> 4099             return coercion_model.bin_op(left, right, mul)
   4100 
   4101         cdef long value

File ~/Desktop/Sage/git/sage/src/sage/structure/coerce.pyx:1222, in sage.structure.coerce.CoercionModel.bin_op()
   1220 if action is not None:
   1221     if (<Action>action)._is_left:
-> 1222         return (<Action>action)._act_(x, y)
   1223     else:
   1224         return (<Action>action)._act_(y, x)

File ~/Desktop/Sage/git/sage/src/sage/matrix/action.pyx:328, in sage.matrix.action.MatrixVectorAction._act_()
    326     else:
    327         v = v.dense_vector()
--> 328 return A._matrix_times_vector_(v)
    329 
    330 

File ~/Desktop/Sage/git/sage/src/sage/matrix/matrix_mod2_dense.pyx:614, in sage.matrix.matrix_mod2_dense.Matrix_mod2_dense._matrix_times_vector_()
    612     return VS.zero()
    613 cdef Vector_mod2_dense c = Vector_mod2_dense.__new__(Vector_mod2_dense)
--> 614 sig_str("matrix allocation failed")
    615 c._init(self._nrows, VS)
    616 if c._entries.nrows and c._entries.ncols:

SignalError: matrix allocation failed
sage: v_ntl * m_ntl
------------------------------------------------------------------------
(no backtrace available)
------------------------------------------------------------------------
Unhandled SIGSEGV: A segmentation fault occurred.
This probably occurred because a *compiled* module has a bug
in it and is not properly wrapped with sig_on(), sig_off().
Python will now terminate.
------------------------------------------------------------------------
[1]    49247 segmentation fault  sage
```
This is a field over which we cannot do linear algebra?

**Comment by tscrim** [comment]
First, you have an input error as `identity_matrix(1, F)` returns a matrix over `ZZ` (which isn't great either), but even swapping the argument order I still get the `SignalError`. There is probably some hidden assumption in the implementation that someone didn't account for different implementations of `GF(q)`. I think there are a number of issues that need to be worked through on a separate PR, but it probably isn't so bad would be my guess.

**Comment by jhpalmieri** [comment]
> First, you have an input error as `identity_matrix(1, F)` returns a matrix over `ZZ` (which isn't great either), but even swapping the argument order I still get the `SignalError`. There is probably some hidden assumption in the implementation that someone didn't account for different implementations of `GF(q)`. I think there are a number of issues that need to be worked through on a separate PR, but it probably isn't so bad would be my guess.

I think that to do it right, someone needs to add doctests for the alternate implementations of finite fields that demonstrate that they can do everything in Sage that the default implementations do. That is a big task, but until it happens, why should anyone trust or expect that `GF(2, impl=...)` will work?

**Comment by jhpalmieri** [comment]
These changes seem to fix this particular problem with `GF(2, impl='ntl')`, but I find them offensive:
```
diff --git a/src/sage/homology/algebraic_topological_model.py b/src/sage/homology/algebraic_topological_model.py
index c7856836f2..9f1148ad6f 100644
--- a/src/sage/homology/algebraic_topological_model.py
+++ b/src/sage/homology/algebraic_topological_model.py
@@ -218,7 +218,7 @@ def algebraic_topological_model(K, base_ring=None):
 
             # c_bar = c - phi(bdry(c))
             c_bar = c_vec
-            bdry_c = diff * c_vec
+            bdry_c = vector(diff * matrix(len(c_vec), 1, c_vec))
             # Apply phi to bdry_c and subtract from c_bar.
             for (idx, coord) in bdry_c.items():
                 try:
@@ -226,7 +226,7 @@ def algebraic_topological_model(K, base_ring=None):
                 except KeyError:
                     pass
 
-            bdry_c_bar = diff * c_bar
+            bdry_c_bar = vector(diff * matrix(len(c_bar), 1, c_bar))
 
             # Evaluate pi(bdry(c_bar)).
             pi_bdry_c_bar = zero
```

**Comment by tscrim** [comment]
> > First, you have an input error as `identity_matrix(1, F)` returns a matrix over `ZZ` (which isn't great either), but even swapping the argument order I still get the `SignalError`. There is probably some hidden assumption in the implementation that someone didn't account for different implementations of `GF(q)`. I think there are a number of issues that need to be worked through on a separate PR, but it probably isn't so bad would be my guess.
> 
> I think that to do it right, someone needs to add doctests for the alternate implementations of finite fields that demonstrate that they can do everything in Sage that the default implementations do. That is a big task, but until it happens, why should anyone trust or expect that `GF(2, impl=...)` will work?

I agree with basically everything you said except the last part. The good news is that it raises an error or crashes rather than return a wrong answer. However, there is more to life than linear algebra (although not much :P): manipulating the fields themselves or polynomials over `GF(p, impl=...)`.

We can also work on it in more piecemeal fashion tool, but hopefully a fix for one will fix the others. At the very least, we can push linear algebra things up to the generic implementation and it really should work there.

**Comment by tscrim** [comment]
> These changes seem to fix this particular problem with `GF(2, impl='ntl')`, but I find them offensive:

I agree, that it's a hack rather than a fix.

For this PR, I think we should keep the code written as is for future-proofing when the other implementations will work. I will leave it up to you what kind of doctests you want for the other implementations.

**Comment by jhpalmieri** [comment]
"The good news is that it raises an error or crashes" — that's the good news. Fun fact: it seems that matrix-vector multiplication with entries in `GF(2, impl='ntl')` actually works if you use sparse matrices. The problems occur with dense matrices. Back when I wrote `algebraic_topological_model.py`, I intentionally used dense matrices over finite fields because according to my tests at the time, they were faster, regardless of sparsity. If sparse matrices had been faster, we would not have found the linear algebra problem because all of this homology stuff would have just worked.

My preference is to not add any doctests for the alternate implementations. Those implementations for `GF(2)` are currently only tested in a few files in `rings/finite_rings`. I expect that once linear algebra works for these implementations, everything else should work. This stuff, or rather the older stuff on which this builds, all relies on linear algebra. The fact that Steenrod operations and cup products work for some simplicial sets suggests that there is nothing intrinsic in the homology code relying on the particular implementation of `GF(2)` beyond linear algebra.

Oh, I just found out that the old stuff doesn't work for all simplicial sets, just for some. Simplicial sets can be somewhat sparser than simplicial complexes, and as a result, the boundary maps in the chain complexes can be zero. I happened to have only tested simplicial sets like that, and when the boundary maps are zero, the algorithm skips the matrix multiplication altogether. I just tried a case with a nontrivial boundary map, and it hits the same problem with matrix multiplication. So even more reason to not include doctests.

I am a little concerned that if we add doctests for `GF(2, impl='ntl')`, it might encourage users to try out these alternate implementations, and that is not likely to go well. If we do add doctests, I would want to add warnings to them, but then someone has to remember to remove the warnings when things have improved. Similarly if we have `TestSuite(...).run(skip=[...])`, we have to remember to go back and delete `skip=[...]` once things work.

In summary: the doctests here wouldn't test anything beyond certain linear algebra computations, and those should be tested elsewhere. There should be nothing in this code that relies on a particular implementation of `GF(2)`. Adding tests could give the wrong impression about the state of `GF(2, impl='ntl')`, and it also creates more work later to remove warnings and add more tests (as things get fixed).

If you really want doctests, I would probably just want to put in failures marked with "# known bug".

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 16]
```diff
@@ -841,6 +1210,194 @@ def Sq(self, i):
...
             return ret
 
+        def _acted_upon_(self, a, self_on_left):
+            r"""
+            Define multiplication of ``self`` by ``a``, an
+            element of the Steenrod algebra.
+
+            INPUT:
+
+            - ``a`` - an element of the mod 2 Steenrod algebra
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
            - ``a`` -- an element of the mod 2 Steenrod algebra
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 17]
```diff
@@ -676,6 +939,95 @@ def cup_product(self, other):
...
+        sage: H = RP4.cohomology_ring(GF(2))
+        sage: x = H.basis()[1,0]
+        sage: Sq(0,1) * x
+        h^{4,0}
+        sage: Sq(3) * x
+        0
+        sage: x * Sq(3)
+        h^{4,0}
+    """
+
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]: We typically do not have the blankline here.

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 18]
```diff
@@ -841,6 +1210,194 @@ def Sq(self, i):
...
+        INPUT:
+
+        - ``deg_domain`` -- the degree of the domain in the cohomology
+          ring
+
+        - ``deg_codomain`` -- the degree of the codomain in the
+          cohomology ring
+
+        - ``side`` default ``'left'``) -- whether we are computing
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
        - ``side`` -- (default ``'left'``) whether we are computing
```

File: src/sage/homology/homology_vector_space_with_basis.py [hunk 19]
```diff
@@ -881,3 +1438,19 @@ def sum_indices(k, i_k_plus_one, S_k_plus_one):
         return [[S_k]]
     return [[i_k] + l for i_k in range(S_k, i_k_plus_one)
             for l in sum_indices(k-1, i_k, S_k)]
+
+def is_GF2(R):
+    """
+    Return ``True`` iff ``R`` is isomorphic to the field GF(2).
```
Comment (tscrim) [inline src/sage/homology/homology_vector_space_with_basis.py]:
```suggestion
    r"""
    Return ``True`` iff ``R`` is isomorphic to the field `\GF{2}`.
```

**Review by tscrim — APPROVED** [review]
Okay, fair enough.

Just a few more little details. Once done, you can set a positive review.

**Comment by jhpalmieri** [comment]
> Okay, fair enough.
> 
> Just a few more little details. Once done, you can set a positive review.

Great, and thank you for all of your feedback!

File: src/sage/homology/homology_vector_space_with_basis.py (hunk 19 above)
Comment (tscrim) [reply src/sage/homology/homology_vector_space_with_basis.py]: It should be `{2}`, not `(2)`. See the output doc.
